v2/tools/generator/internal/conversions/README.md

```diff
@@ -0,0 +1,568 @@
+            // reading a local variable or a field
+            cacheOriginal = nil
+            actualReader = reader
+        default:
+            // Something else, so we cache the original
+            local := knownLocals.CreateSingularLocal(
+                sourceEndpoint.Name(), 
+                "", 
+                "Cache")
+            cacheOriginal = astbuilder.ShortDeclaration(
+                local, 
+                reader)
+            actualReader = dst.NewIdent(local)
+        }
+```
+
+* `cacheOriginal` is an optional statement that reads our value and stashes it in a local variable.
+* `actualReader` is an expression for use later in the function. It reads from our local variable (if used), or directly from our source (if not).
+
+Our array may be nil, so we need to check and avoid dereferencing it if so.
+
+``` go
+        checkForNil := astbuilder.AreNotEqual(actualReader, astbuilder.Nil())
+```
+
+Using `knownLocals`, we create three obviously related identifiers to use for the array conversion. Each uses the name of our source as a base, with a different suffix.
+
+We use different suffixes here to those used in `assignMapFromMap` in order to keep the code easy to read if the two are nested.
+
+``` go
+        branchLocals := knownLocals.Clone()
+        tempId := branchLocals.
+            CreateSingularLocal(sourceEndpoint.Name(), "List")
+
+        loopLocals := branchLocals.Clone()
+        itemId := loopLocals.
+            CreateSingularLocal(sourceEndpoint.Name(), "Item")
+        indexId := loopLocals.
+            CreateSingularLocal(sourceEndpoint.Name(), "Index")
+```
+
+* `List` is the final list we're creating
+* `Item` is a single item from the source list
+* `Index` is the integer index of the item we're converting
+
+Clones of `knownLocals` are used for each nested scope (one inside our **if** statement, see below; the other inside our loop) so that definitions don't leak out.
+
+Our declaration initializes a new list of the correct final size.
+
+``` go
+        declaration := astbuilder.ShortDeclaration(
+            tempId,
+            astbuilder.MakeSlice(
+                destinationArray.AsType(generationContext), 
+                astbuilder.CallFunc("len", actualReader)))
+```
+
+We create a custom writer to store values into our at a given index.
+
+``` go
+        writeToElement := func(expr dst.Expr) []dst.Stmt {
+            return []dst.Stmt{
+                astbuilder.SimpleAssignment(
+                    &dst.IndexExpr{
+                        X:     dst.NewIdent(tempId),
+                        Index: dst.NewIdent(indexId),
+                    },
+                    expr),
+            }
+        }
+```
+
+Within our loop we create a copy of the item variable, to avoid aliasing between loops. We know for our sample case that an alias isn't necessary - but we don't know generally know what an conversions might do, so this is a safety precaution.
```

Comment:
```suggestion
Within our loop we create a copy of the item variable, to avoid aliasing between loops. We know for our sample case that an alias isn't necessary - but we don't know generally know what conversions might do, so this is a safety precaution.
```